### PR TITLE
Health monitor and metrics collection

### PR DESCRIPTION
**Pull request by eatstoomuchjam**
Add a basic health monitor and report per-database statistics from postgres' built-in stats database.

### CODE REVIEW

File: layer-build.image [hunk 1]
```diff
@@ -1 +1 @@
-registry.infra.layer.com/vendor-golang:1.0.0
+registry.infra.layer.com/vendor-golang:ae76755179a7f85a35177939e5feeb75a48a6b28
```
Comment (jessedavidson) [inline layer-build.image]: Should be by tag

File: postgresql-health/postgresql-health.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		responseCode = 503
+	}
+
+	if r.Method == "HEAD" {
+		w.WriteHeader(responseCode)
+		return
+	}
+
+	serviceMetric := &servicemetric.ServiceMetric{
+		Service: "postgresql",
```
Comment (jessedavidson) [inline postgresql-health/postgresql-health.go]: This should come from the environment variable SERVICE_POOL, which will come from https://github.com/layerhq/deploy/blob/master/templates/postgresql.service.tmpl

Matters here because we're going to have multiple postgresql templates

**Comment by jessedavidson** [comment]
Let's avoid merge commits please

**Comment by eatstoomuchjam** [comment]
Sorry, the merge commit stays.  They are a reality of someone having done non-trivial work on a branch when you do a commit to master.

File: postgresql-health/postgresql-health.go (hunk 2 above)
Comment (eatstoomuchjam) [inline postgresql-health/postgresql-health.go]: I have a branch for laygo to add a service pool function to it.  As soon as that's merged, I'll update this.

**Comment by jessedavidson** [comment]
I announced this work yesterday. You knew it was happening.

This is not the first time we've talked about merge commits, and shouldn't come as a surprise to you.

**Comment by eatstoomuchjam** [comment]
Yeah.  I knew you were updating master and that I'd end up having to do a merge since my branch was already pretty far along when you did it.  That shouldn't be a blocker for you.

Complex git surgery is not an acceptable to me. I humored you on it once before.  I'm meeting you halfway by heavily rebasing my branches even though I strongly dislike doing it.  One of the hazards of having multiple branches on a shared repo is that there might be merge commits sometimes.

**Comment by jessedavidson** [comment]
My apologies for giving you the impression that this is a personal preference.

It's a Layer Engineering best practice. It's not documented, yet; apologies for that as well. But we do avoid merge commits from master into branches. They make the history harder to read.

The client and server teams have been holding to this standard for a long time. It hasn't come up for us because we have fewer people and more repositories. But now that it has come up, I would like infrastructure to maintain the same standard.

Please make the change.

**Comment by eatstoomuchjam** [comment]
Any git workflow which includes frequent use of force pushes and history rewriting is not a "best practice."  It's a behavior that scales extremely poorly with team growth and ends up (in general) causing more problems than it solves.  You would like infrastructure to adopt that standard, but I would like infrastructure not to make the same mistakes.
I've chatted about this with a number of friends who use git at work.  Most have never even tried that sort of workflow and those who have ended up giving up on it because it resulted in too much time wasted dealing with git problems and not in solving the problems they needed solved.

**Comment by blakewatters** [comment]
Arbitrarily merging branches back and forth makes the history extremely hard to read and prevents the assignment of any clear roles and semantics to the branches. In order to have a sane, readable history it’s imperative that branches are merged in one direction. Merges should mean something — the bringing together of parallel lines of development to form a new joined line of development. `master` is the ultimate source of truth branch and should only receive `--no-ff` merges to clearly demarcate when a new branch has rejoined the mainline.

Can you be specific about the problems that you perceive as being imminent that you are trying to save us from? All the previous pushback I’ve ever heard of on these points doesn’t sustain being held up to the light. Rebasing is a fundamental part of life with git — not something to be avoided at all. Using the available tools like utilizing smaller, short-lived branches, having single owners for features, practicing squashing and enabling [git rerere](http://git-scm.com/docs/git-rerere) can mitigate much of the pain of rebasing feature branches.

The idea that this sort of workflow doesn’t scale as the number of developers grows just doesn’t hold water. Take a look at the history of the master branch on the [Linux Kernel Git Mirror](https://github.com/torvalds/linux/commits/master). There’s not a single merge of code from master back into itself as the result of merging a feature branch or patchset. Closer to home, the client SDK’s (which by the way represent the largest codebases within Layer) have maintained feature branches of 27k net line changes for 3+ months that were rebased against master on an on-going basis.

IMHO this line of argument is similar to “We don’t write tests because they take too long”. It’s a false economy.

I’ve long advocated that we as an engineering organization officially adopted the [Git Flow](http://nvie.com/posts/a-successful-git-branching-model/) model (and associated tooling) specifically to prevent threads like this.

**Comment by eatstoomuchjam** [comment]
@blakewatters I completely disagree with the statement that "It is imperative that branches be merged in one direction."  It is absolutely not imperative that be the case and in my most recent role I worked on a codebase with between 10 and 20 other developers contributing to it constantly.  On the rare occasions that it was difficult to track down when something changed in the past, the culprit was never a merge commit - not once in three years, and it rarely took us more than a few minutes to track it down.

Some of the stuff you've mentioned, including "having single owners for branches" sounds great in an ideal world, but doesn't always happen in the real world.  Sometimes there are multiple owners for branches.  I wasn't familiar with `git rerere` before.  It looks neat and it would certainly reduce inconvenience a bit.

Another problem with having a workflow where force pushes are routine and everybody has access to master is that it just takes one accidental slip to obliterate commit history for master in a way that can be difficult to restore (you'd better hope someone else pulled the latest changes!).  
There was an instance shortly after I started where Jesse decided something shouldn't be in master so he blew away the commits instead of just using revert commits.  Whatever he did broke my local copy to the point where a hard reset didn't work and I ended up re-cloning the repo to avoid wasting my day.

Basically, the main reason I would assert that it doesn't scale well is that humans are fallible and it's fundamentally broken to create policies which allow a routine part of the daily workflow be a dangerous action that can massively inconvenience everybody else.

**Comment by blakewatters** [comment]
Jesse should be flogged for force pushing master. A corollary to the workflow outlined above is that you **never** force push integration branches.

And if you have more than one person working on a branch you can defer rebasing until its ready for integration/code review or coordinate the transition by rebasing and pushing a derived branch onto a new name. Presumably if you are working with another peer they aren’t going appreciate having everything from master code-fisted into your shared workspace anyway.

Aside from blindly force pushing integration branches any sort of trouble you can possibly get yourself in can be immediately unwound by leveraging the existing Git functionality. Rebase comes with a `--abort` argument that can bail out of the operation and any history rewriting that you do complete can be recovered by utilizing `git reflog`.

I’ve done a ton of work on large projects both open and closed source and have never encountered the “merge whatever, whenever the fuck you want” workflow before.

In any case, you are going to have to get with the program because we have established engineering processes and practices that are not on an opt-in basis. If you want to put forth a formal proposal to make changes to that process then that is fine and will be considered by the team, but you cannot unilaterally decide to ignore it.
